File: Java/com/gildedrose/GildedRose.java
package com.gildedrose;

import static com.gildedrose.GildedRose.Ageing.NoAgeing;
import static com.gildedrose.GildedRose.Ageing.StandardAgeing;
import static com.gildedrose.GildedRose.Degradation.*;
import static com.gildedrose.GildedRose.Saturation.NoSaturation;

class GildedRose {
    Item[] items;

    public GildedRose(Item[] items) {
        this.items = items;
    }

    public void updateQuality() {
        for (Item item : items) {
            update(item);
        }
    }

    void update(Item item) {
        typeOf(item).update(item);
    }

    private ItemType typeOf(Item item) {
        String name = item.name;
        if (name.equals("Sulfuras, Hand of Ragnaros"))
            return ItemType.Sulfuras;
        if (name.equals("Aged Brie"))
            return ItemType.Brie;
        if (name.equals("Backstage passes to a TAFKAL80ETC concert"))
            return ItemType.Pass;
        if (name.startsWith("Conjured"))
            return ItemType.Conjured;
        return ItemType.Other;
    }

    enum ItemType {
        Sulfuras(NoAgeing, NoDegradation, NoSaturation),
        Brie(StandardAgeing, BetterWithTime, Saturation.StandardSaturation),
        Pass(StandardAgeing, PassDegradation, Saturation.StandardSaturation),
        Conjured(StandardAgeing, ConjuredDegragation, Saturation.StandardSaturation),
        Other(StandardAgeing, StandardSaturation, Saturation.StandardSaturation);

        public void update(Item item) {
            ageing.apply(item);
            degradation.apply(item);
            saturation.apply(item);
        }

        private final Ageing ageing;
        private final Degradation degradation;
        private final Saturation saturation;

        ItemType(Ageing ageing, Degradation degradation, Saturation saturation) {
            this.ageing = ageing;
            this.degradation = degradation;
            this.saturation = saturation;
        }
    }

    enum Ageing {
        StandardAgeing(1),
        NoAgeing(0);

        public void apply(Item item) {
            item.sellIn = item.sellIn - sellInChangePerDay;
        }

        private final int sellInChangePerDay;

        Ageing(int sellInChangePerDay) {
            this.sellInChangePerDay = sellInChangePerDay;
        }
    }

    enum Degradation {
        NoDegradation,
        BetterWithTime {
            public void apply(Item item) {
                item.quality = item.quality + 1;
                if (pastSellBy(item)) {
                    item.quality = item.quality + 1;
                }
            }
        },
        PassDegradation {
            public void apply(Item item) {
                item.quality = item.quality + 1;

                if (item.sellIn < 10) {
                    item.quality = item.quality + 1;
                }

                if (item.sellIn < 5) {
                    item.quality = item.quality + 1;
                }

                if (pastSellBy(item))
                    item.quality = 0;
            }
        },
        StandardSaturation {
            public void apply(Item item) {
                item.quality = item.quality - 1;
                if (pastSellBy(item))
                    item.quality = item.quality - 1;
            }
        },
        ConjuredDegragation {
            public void apply(Item item) {
                item.quality = item.quality - 2;
                if (pastSellBy(item))
                    item.quality = item.quality - 2;
            }

        };

        private static boolean pastSellBy(Item item) {
            return item.sellIn < 0;
        }

        public void apply(Item item) {}
    }

    enum Saturation {
        StandardSaturation {
            public void apply(Item item) {
                if (item.quality > 50)
                    item.quality = 50;
                if (item.quality < 0)
                    item.quality = 0;
            }
        },
        NoSaturation;

        public void apply(Item item) {}
    }
}

File: Kotlin/src/main/kotlin/com/gildedrose/GildedRose.kt
package com.gildedrose

import kotlin.math.max

class GildedRose(
    private val items: List<Item>,
) {
    fun updateQuality() {
        for (item in items) {
            item.type().update(item)
        }
    }
}

fun Item.type() = when (name) {
    "Sulfuras, Hand of Ragnaros" -> Sulfuras
    "Aged Brie" -> Brie
    "Backstage passes to a TAFKAL80ETC concert" -> Passes
    "Conjured Mana Cake" -> ConjuredCake
    else -> Other
}

open class ItemType {

    fun update(item: Item) {
        item.sellIn -= ageing()
        item.quality = (item.quality - degradationFor(item))
            .coerceIn(0, max(50, item.quality))
    }

    protected open fun ageing() = 1

    protected open fun degradationFor(item: Item) = if (item.sellIn < 0) 2 else 1
}

object Sulfuras : ItemType() {
    override fun ageing() = 0
    override fun degradationFor(item: Item) = 0
}

object Brie : ItemType() {
    override fun degradationFor(item: Item) = if (item.sellIn < 0) -2 else -1
}

object Passes : ItemType() {
    override fun degradationFor(item: Item) = when {
        item.sellIn < 0 -> item.quality
        item.sellIn < 5 -> -3
        item.sellIn < 10 -> -2
        else -> -1
    }
}

object ConjuredCake : ItemType() {
    override fun degradationFor(item: Item) = 2 * super.degradationFor(item)
}

object Other : ItemType()